https://github.com/eblot/pyftdi/pull/356
https://github.com/eblot/pyftdi/commit/c131ff2c302b1bfe2ec6ec53331f55cae16b4b66.patch

From b82c5e264e82052440cf98da070d86f81c734c4a Mon Sep 17 00:00:00 2001
From: Mark Mentovai <mark@mentovai.com>
Date: Wed, 16 Aug 2023 12:00:04 -0400
Subject: [PATCH 1/2] tests: use module_setup to establish virtual mock
 environment

This makes it possible to run these virtual mock tests under pytest,
which does not run main.
---
 pyftdi/tests/eeprom_mock.py | 12 ++++++------
 pyftdi/tests/gpio.py        | 12 ++++++------
 pyftdi/tests/mockusb.py     |  6 +++---
 3 files changed, 15 insertions(+), 15 deletions(-)

diff --git pyftdi/tests/eeprom_mock.py pyftdi/tests/eeprom_mock.py
index 7acf99f3048d..78665789c4c8 100644
--- pyftdi/tests/eeprom_mock.py
+++ pyftdi/tests/eeprom_mock.py
@@ -269,7 +269,7 @@ def virtualize():
         raise AssertionError('Cannot load virtual USB backend')
 
 
-def main():
+def setup_module():
     import doctest
     doctest.testmod(modules[__name__])
     debug = to_bool(environ.get('FTDI_DEBUG', 'off'))
@@ -288,10 +288,6 @@ def main():
     FtdiLogger.set_level(loglevel)
     FtdiLogger.set_formatter(formatter)
     virtualize()
-    try:
-        ut_main(defaultTest='suite')
-    except KeyboardInterrupt:
-        pass
 
 
 if __name__ == '__main__':
@@ -300,4 +296,8 @@ if __name__ == '__main__':
     #  FTDI_LOGLEVEL: a Logger debug level, to define log verbosity
     #  FTDI_DEBUG: to enable/disable debug mode
     #  FTDI_VIRTUAL: to use a virtual device rather than a physical device
-    main()
+    setup_module()
+    try:
+        ut_main(defaultTest='suite')
+    except KeyboardInterrupt:
+        pass
diff --git pyftdi/tests/gpio.py pyftdi/tests/gpio.py
index 4d159d51ac78..bdb0cbc892ea 100755
--- pyftdi/tests/gpio.py
+++ pyftdi/tests/gpio.py
@@ -633,7 +633,7 @@ def virtualize():
         raise AssertionError('Cannot load virtual USB backend') from exc
 
 
-def main():
+def setup_module():
     import doctest
     doctest.testmod(modules[__name__])
     debug = to_bool(environ.get('FTDI_DEBUG', 'off'))
@@ -652,10 +652,6 @@ def main():
     FtdiLogger.set_level(loglevel)
     FtdiLogger.set_formatter(formatter)
     virtualize()
-    try:
-        ut_main(defaultTest='suite')
-    except KeyboardInterrupt:
-        pass
 
 
 if __name__ == '__main__':
@@ -664,4 +660,8 @@ if __name__ == '__main__':
     #  FTDI_LOGLEVEL: a Logger debug level, to define log verbosity
     #  FTDI_DEBUG: to enable/disable debug mode
     #  FTDI_VIRTUAL: to use a virtual device rather than a physical device
-    main()
+    setup_module()
+    try:
+        ut_main(defaultTest='suite')
+    except KeyboardInterrupt:
+        pass
diff --git pyftdi/tests/mockusb.py pyftdi/tests/mockusb.py
index 57d03e46d2cc..17553b1e7d88 100755
--- pyftdi/tests/mockusb.py
+++ pyftdi/tests/mockusb.py
@@ -839,7 +839,7 @@ def suite():
     return suite_
 
 
-def main():
+def setup_module():
     testmod(modules[__name__])
     debug = to_bool(environ.get('FTDI_DEBUG', 'off'))
     if debug:
@@ -866,8 +866,8 @@ def main():
         MockLoader = backend.create_loader()
     except AttributeError as exc:
         raise AssertionError('Cannot load virtual USB backend') from exc
-    ut_main(defaultTest='suite')
 
 
 if __name__ == '__main__':
-    main()
+    setup_module()
+    ut_main(defaultTest='suite')
-- 
2.41.0

